### PR TITLE
New parser fixes

### PR DESCRIPTION
**Pull request by wdanilo**
### Pull Request Description

This pull request improves several things:
- Expression `"` is now parsed correctly
- Multiline strings are handled better
- There are several bugfixes including wrong codespan computation when providing explicit multiline constructors.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] The code conforms to the [Luna Style Guide](https://github.com/luna/wiki/blob/master/code-style/01.general.md).
- [x] The code has been tested where possible.

### CODE REVIEW

File: core/src/Luna/IR/Term/Ast/Invalid.hs [hunk 1]
```diff
@@ -92,3 +92,7 @@ emptyExpression = convert' EmptyExpression
 missingSection :: Convertible' Symbol a => a
 missingSection = convert' MissingSection
 {-# INLINE missingSection #-}
+
+stringNoClosingMark :: Convertible' Symbol a => a
+stringNoClosingMark = convert' (Literal $ String NoClosingMark)
```
Comment (ljkania) [inline core/src/Luna/IR/Term/Ast/Invalid.hs]: `convert' . Literal $ String NoClosingMark` works as well

File: lib/parser-utils/src/Text/Parser/State/Indent.hs [hunk 2]
```diff
@@ -66,29 +66,22 @@ withRoot :: State.Monad Indent m => m a -> m a
...
 {-# INLINE guard #-}
 
 indented, indentedOrEq, indentedEq
-    :: State.Getters '[Indent, Position] m => m ()
-indented     = guard (== GT) "Expected indentation"
-indentedOrEq = guard (/= LT) "Expected indentation"
-indentedEq   = guard (== EQ) "Indentation does not match the previous one"
+    :: State.Getters '[Indent, Position] m => m Delta
+indented     = guard (>  0) "Expected indentation"
+indentedOrEq = guard (>= 0) "Expected indentation"
```
Comment (ljkania) [inline lib/parser-utils/src/Text/Parser/State/Indent.hs]: Message repetition, consider refactor to constant.

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs [hunk 3]
```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
 strBuilder quote = TokenStream.add =<< spanned parser where
     parser = do
         let isQuote      = (== quote)
-            isBodyChar c = c == quote || isEolBeginChar c
+            isBodyChar c = not $ c == quote || isEolBeginChar c
```
Comment (ljkania) [inline syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: You've defined `isQuote` line before - why don't you use it?

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs [hunk 4]
```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
...
+            body         = Ast.Str <$> chunks
+        if quoteLen == 2 then pure (Ast.Str mempty) else (body <* end) <|> (Ast.Str mempty <$ noEnd)
 
 
 --------------------
 -- === Layout === --
 --------------------
 
-flexBlock1 :: Lexer a -> Lexer (NonEmpty a)
+strBlock1 :: forall t. Lexer (Spanned (Atom.StrChunk t)) -> Lexer [Spanned (Atom.StrChunk t)]
```
Comment (ljkania) [inline syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: Shouldn't this name be more descriptive?

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs [hunk 5]
```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
...
+strBlock1 = \p -> let
+    line :: Bool -> Lexer (Spanned Delta, [Spanned (Atom.StrChunk t)])
+    line isFirst = do
+        off <- spanned $ newline *> Indent.indented
+        (off,) <$> many p
```
Comment (ljkania) [inline syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: `isFirst` is never used.

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs [hunk 6]
```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
...
+strBlock = option mempty . strBlock1
+{-# INLINE strBlock #-}
+
+
+mergeStrChunks :: [Spanned (Atom.StrChunk t)] -> [Spanned (Atom.StrChunk t)]
+mergeStrChunks = \case
+    []     -> []
+    (a:as) -> let
+        span   = view Ast.span a
+        spans  = view Ast.span <$> as
```
Comment (ljkania) [inline syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: There is a code repetition, consider refactor.

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs [hunk 7]
```diff
@@ -663,33 +664,77 @@ strBuilder :: Char -> Lexer_
...
+        getTxt = \(Atom.StrPlain t) -> t
+        txt    = getTxt . Ast.unsafeUnspan  $  a
+        txts   = getTxt . Ast.unsafeUnspan <$> as
```
Comment (ljkania) [inline syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: ditto.

**Comment by mikusp** [comment]
Rebase this branch or create a new one since >110 commits from this pull request are already merged into master

**Comment by mikusp** [comment]
Also, CI says that tests fail so fix that beforehand.

**Comment by piotrMocz** [comment]
@wdanilo what's the status of this pull request? Will you be fixing the tests and merging it?

**Comment by iamrecursion** [comment]
The tests definitely need fixing. If my review is _required_ please let me know. I'd also like to see it rebased.

**Comment by wdanilo** [comment]
Definitely. I'll try to do it today. If I would not be able to, I'll do it max tomorrow in the office.

**Comment by iamrecursion** [comment]
This definitely needs to land soon.

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs (hunk 4 above)
Comment (wdanilo) [reply syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: It's a block parser for string literals. It's a rather standard name for parser-like utils. Leaving as is for now, but if there are propositions for better names, I'd love to revisit this topic.

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs (hunk 6 above)
Comment (wdanilo) [reply syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: This is just a lens accessor, it's not a code repetition.

File: syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs (hunk 5 above)
Comment (wdanilo) [reply syntax/text/parser2/src/Luna/Syntax/Text/Parser/Lexer/Grammar.hs]: fixed, good catch

**Comment by wdanilo** [comment]
Ready to be merged. If someone knows how to rebase it as @mikusp proposed, please do it. @ljkania comments applied, tests working.

**Pull request merged**